File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <string>
#include <vector>
#include <stack>
#include <queue>
#include <iostream>

using namespace std;
vector<pair<int, int>> pairs = {{'(', ')'}, {'[', ']'}, {'{', '}'}};
bool check(queue<char> q){ // 그냥 복사해오기
    stack<char> s; 
    int size = q.size();
    for(int i = 0; i < size; i++){
        // 여는거면 넣고, 빼는거면 제거
        char qf = q.front(); q.pop();
        // 여는 괄호라면
        if(qf == '(' || qf == '[' || qf == '{') {
            s.push(qf);
        } 
        // 닫는 괄호라면
        else{
            for(int i = 0; i < 3; i++){
                if(qf == pairs[i].second){
                    if(s.empty()) { return false;}
                    else if (s.top() == pairs[i].first) { s.pop();}
                    else {s.push(qf);} 
                }
            }
        }
    }
    if(s.empty()) return true;
    else return false;
}

int solution(string s) {
    int answer = 0;
    int size = s.size();
    
    // 큐에 넣기
    queue<char> q;
    for(int i = 0; i < size; i++){
        q.push(s[i]);
    }
    if(check(q)) answer++;
    
    // 회전시키고 확인 반복
    for(int i = 0; i < size-1; i++){
        q.push(q.front());
        q.pop();
        if(check(q)) answer++;
    }

    return answer;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*; 

class Solution {
    
    Deque<Character> stack = new ArrayDeque<>(); 
    
    public int solution(String s) {
        int answer = 0; 
        
        StringBuilder sb = new StringBuilder(); 
        sb.append(s);
        if(isAvail(sb)) answer++; 
        // 회전 시키고 확인하고 반복 - 1,000,000
        for(int i = 0; i < s.length() -1; i++){
            char first = sb.charAt(0); 
            sb.append(first); 
            sb.delete(0, 1); 
            if(isAvail(sb)) answer++; 
        }
        
        // s를 왼쪽으로 x만큼 회전시켰을때 x가 올바른 괄호 문자열이 되게하는 x의 개수를 리턴
        return answer;
    }
    
    public boolean isAvail(StringBuilder sb){
        stack.clear(); 
        for(int i = 0; i < sb.length(); i++){
            char cur = sb.charAt(i); 
            // 제거
            if(!stack.isEmpty()){
                if(cur == ')' && stack.peek() == '(') stack.pop();
                else if(cur == '}' && stack.peek() == '{') stack.pop();
                else if (cur == ']' && stack.peek() == '[') stack.pop();
                else stack.push(cur); 
            }
            // 담기
            else stack.push(cur); 
        }
        if(stack.isEmpty()) return true; 
        else return false; 
    }
}